### PR TITLE
Add support for the K70 RGB TKL

### LINKED ISSUE #745: Corsair K70 RGB TLK
Hi there,

The keyboard K70 RGB TLK are not supported by ckb-next. No device is detected.
It's possible for you to implement it ?

Thank you for your work

---

* Version ckb-next : 0.4.4-99-g761ba48
* Report ckb-next : [ckb-next-report-1631657442.gz](https://github.com/ckb-next/ckb-next/files/7165434/ckb-next-report-1631657442.gz)

--- 
* Keyboard Name : Corsair K70 RGB TLK Mechanical Gaming Keyboard
* Keyboard Model : RGP0071
* Official Keyboard Web page : https://www.corsair.com/us/en/Categories/Products/Gaming-Keyboards/RGB-Mechanical-Gaming-Keyboards/K70-RGB-TKL-CHAMPION-SERIES-Mechanical-Gaming-Keyboard/p/CH-9119010-NA

### PR DESCRIPTION
**Pull request by minneyar**
This adds support for the K70 RGB TKL Champion Series keyboard.  Product page is here: https://www.corsair.com/us/en/Categories/Products/Gaming-Keyboards/RGB-Mechanical-Gaming-Keyboards/K70-RGB-TKL-CHAMPION-SERIES-Mechanical-Gaming-Keyboard/p/CH-9119014-NA

Its interface is fairly similar to the K100, but its key layout is closer to the K63.

I've tested this with my own keyboard, and everything seems fully functional.

Fixes #745

### CODE REVIEW

**Comment by thekhalifa** [comment]
Compiled and tested your branch on mine as well and seems fully functional.
I thought the lock/mute keys were broken, but with ckb running they behave differently to hardware mode.

Thanks for patching this.

**Comment by minneyar** [comment]
Thanks for testing it!  That did also confuse me a little at first.  It's possible to customize the behavior of the brightness/lock/mute keys in the ckb-next gui; they're under the "Performance" tab in the keyboard settings.  The default behavior is a little different from hardware mode.

File: src/gui/kbperf.cpp [hunk 1]
```diff
@@ -488,9 +488,11 @@ void KbPerf::applyIndicators(int modeIndex, const bool indicatorState[]){
         lightIndicator("dpi", dpiClr[index].rgba());
     }
     // KB indicators
-    // Disable the M indicators for the K70MK2 and the STRAFE_MK2.
+    // Disable the M indicators for the K70MK2, the STRAFE_MK2, and the K70_TKL.
     // FIXME: Only enable them for devices that need them instead
-    if(iEnable[MODE] && !(this->modeParent()->bind()->map().model() == KeyMap::K70MK2 || this->modeParent()->bind()->map().model() == KeyMap::STRAFE_MK2)){
+    if(iEnable[MODE] && !(this->modeParent()->bind()->map().model() == KeyMap::K70MK2 ||
+                          this->modeParent()->bind()->map().model() == KeyMap::STRAFE_MK2 ||
+                          this->modeParent()->bind()->map().model() == KeyMap::K70_TKL)){
```
Comment (tatokis) [inline src/gui/kbperf.cpp:495]: Are you sure this is needed?

File: src/gui/keymap.cpp [hunk 2]
```diff
@@ -732,6 +743,29 @@ static QHash<QString, Key> getMap(KeyMap::Model model, KeyMap::Layout layout){
         map["voldn"].x -= 10;
         break;
     }
+    case KeyMap::K70_TKL:{
+        map = getMap(KeyMap::K70, layout);
+        // Very similar to the K63 but with slightly different media button positions
```
Comment (tatokis) [inline src/gui/keymap.cpp:748]: If the only difference between the K63 and the K70TKL is just the top row, you can take the K63 map and then patch the top row on top if it. Should reduce duplicate code

**Review by tatokis — CHANGES_REQUESTED** [review]
Apologies. I had started the review a few days after you opened the PR but completely forgot about it and didn't submit it. I can't remember what else I wanted to bring up, but for now this will do.

**Comment by gtjoseph** [comment]
Any progress on this?  @minneyar Are you going to be able to finish?

**Comment by avanwyk** [comment]
My knowledge of C/C++ is limited, but I could try to take a run at this.

**Comment by gardotd426** [comment]
Is this dead?

**Comment by Hotshot0592** [comment]
Hey there, new to linux and this site. Please bare with me if I step on any toes in regards to ignorant questions. But, is there any hope of when this patch might be launched or updated?

**Comment by minneyar** [comment]
I haven't had time to look at this for quite a while and it works as-is, so I had forgotten about it, but I'll see if I can clean it up a bit to bring it up to date with the target branch and address the requested change.

On the other hand, the branch it's based on hasn't been updated since August last year, so I don't know if there are any plans to merge it into the master branch and release it any time soon.

For what it's worth, this branch still works fine as-is if you build and install it yourself.

**Comment by Hotshot0592** [comment]
Still learning coding. So is there a thing for now where I could copy paste
into my terminal? And I only know python.

**Comment by Hotshot0592** [comment]
Or would it be better/easier for someone at my level to just go buy a new
compatible keyboard? I was thinking about doing so within the next few
months anyways.

**Comment by minneyar** [comment]
@tatokis Sorry for the (very) long delay on this; I cleaned up the above comments and brought this up to date with the target branch, and it looks like everything still works for me.  Let me know if there are any other issues.

@Hotshot0592 Building and installing it should be pretty straightforward; you'll want to clone my repo and checkout the `k70tkl` branch, then follow these instructions for building from source: https://github.com/ckb-next/ckb-next/wiki/Linux-Installation#build-from-source

**Comment by Hotshot0592** [comment]
Yeah, I copied it in python. Do I just throw it in terminal now?

**Comment by minneyar** [comment]
> Yeah, I copied it in python. Do I just throw it in terminal now?

There's no need to do anything in Python, the commands in the wiki are what you'd run in a terminal to clone and build the repository; although GitHub successfully built packages from my branch, so you can probably just download the artifacts here and install the right package for your platform: https://github.com/minneyar/ckb-next/actions/runs/4970360775

**Comment by Hotshot0592** [comment]
Now I have ckb-next installed already. Do I just update it somehow?

**Comment by Hotshot0592** [comment]
Again, sorry. Completely new.

**Comment by minneyar** [comment]
> Now I have ckb-next installed already. Do I just update it somehow?

Normally installing a newer version of a package will automatically update it, but it might not in this case since the version number in my branch hasn't changed from the source branch.  It's probably easiest to uninstall the ckb-next package you had previous installed, then install the one built from my repo, although exact instructions on how to do that could vary depending on what Linux distro you're using.

**Comment by Hotshot0592** [comment]
So I keep getting an error when downloading your package.
image_2023-05-14_084127787

**Comment by Hotshot0592** [comment]
I'm using Nobara, which is derived from Fedora if that helps at all. Also, I tried both of the Fedora packages you have on there.

**Comment by minneyar** [comment]
I think Nobara is based on Fedora 37, which is newer than any of the packages that were being generated by that build.  I added Fedora 37 and rebuilt it, so try the `ckb-next.fedora-37.GNU.Release.rpm` package from this one: https://github.com/minneyar/ckb-next/actions/runs/4972785546

**Comment by Hotshot0592** [comment]
I tried and this is what I got.
Screenshot from 2023-05-14 09-13-26

**Comment by minneyar** [comment]
Hmm.  Unfortunately, I'm not really very familiar with Red Hat or its packaging system enough to debug why it doesn't like that; at this point it's probably easier to try building it from source.

For reference, the commands you'd run in a terminal to do that but using my fork and branch instead of this repo would look like:

1. Install prerequisites
   ```
   sudo dnf install gcc gcc-c++ make cmake glibc zlib-devel qt5-qtbase-devel quazip-qt5-devel systemd-devel pulseaudio-libs-devel qt5-linguist qt5-qtx11extras-devel xcb-util-wm-devel xcb-util-devel libxcb-devel git dbusmenu-qt5-devel
   ```
2. Clone the repo, cd into it, checkout the branch:
   ```
   git clone https://github.com/minneyar/ckb-next.git
   cd ckb-next
   git checkout k70tkl
   ```
3. Run the script to build and install it:
   ```
   ./quickinstall
   ```

**Comment by Hotshot0592** [comment]
I got it working, thanks! Now I have been doing all of this to see if one program will work. It is call STO Keybind, it is for Star Trek Online. It works just fine on windows (shocker right) and now when I go to select what key I want to do what with it pulls up on a blank drop down. Everything up to that point works. Any idea where I can go to post about this? I thought it needed drivers for it to work and detect my keyboard. But Idk if that is still the case or not.

**Comment by Hotshot0592** [comment]
Could part of the issue be that I am getting an error when I try updating my flatpak? But I cannot update the cbk-next due to some error saying that the directory does not exist.

**Comment by tatokis** [comment]
@minneyar Apologies for the late reply. I ended up writing some code to simplify keymap patching (which means there's no need to manually delete patched keymap entries anymore).

I made some minor changes and rebased against master (which I merged the k100-v2 branch into earlier). Keep in mind I force pushed.

Please verify that everything still functions as expected, and if so, I'll merge it.

**Comment by HiWhatName** [comment]
Can confirm, it's fully functional.

**Comment by minneyar** [comment]
Yep, I just tried it out and it looks good to me.

**Comment by MysteriousRSA** [comment]
Has there been any updates on this?

When i install ckb-next it doesnt show still and is listed as requested in the supported devices, same with... all of my other peripherals

**Comment by enigamus** [comment]
> Has there been any updates on this?
> 
> When i install ckb-next it doesnt show still and is listed as requested in the supported devices, same with... all of my other peripherals

This was about K70 RGB TKL Mechanical, you may have the K70 RGB TKL Optical-Mechanical Keyboard (with Corsair Opx). 
If you run `ckb-next-dev-detect` and look for idProduct, "0x1b73" is supported, "0x1bb9" is not.

There is an [open  issue](https://github.com/ckb-next/ckb-next/issues/1003).

**Comment by discotimetraveler** [comment]
I just found this software and this post. I have the K70 TKL Mechanical version and I installed from apt-get (0.6.0) and my keyboard is not detected...? (My mouse is detected in the software)....   my -dev-detect...

idProduct          0x1b73 CORSAIR K70 RGB TKL CHAMPION SERIES Mechanical Gaming Keyboard

$ ckb-next -v
ckb-next 0.6.0

I have the latest bios. I was using the same keyboard on WIn11 on the same system. It's an Intel NUC so not a lot of USB ports. I went through the basic TS steps; other ports, reboot, unplug/plug, I don't think this keyboard has a BIOS switch (just this tournament switch)....  Any suggestions?

Running Debian (Sparky 8) with 6.6.11 kernel.

Thanks!

**Pull request merged**